pkg/broker/broker.go

```diff
@@ -850,10 +859,194 @@ func (a AnsibleBroker) Unbind(
+	   * async provision 202 Accepted
+	   * if only support async and no accepts_incomplete=true passed in, 422 Unprocessable entity
+
+	*/
+
+	var err error
+
+	si, err := a.dao.GetServiceInstance(instanceUUID.String())
+	if err != nil {
+		a.log.Debug("Error retrieving instance")
+		return nil, ErrorNotFound
+	}
+
+	// If no plan was specified we'll continue with the current plan
+	if req.PlanID == "" {
+		req.PlanID = (*si.Parameters)["_apb_plan_id"].(string)
+	}
+
+	// Retrieve requested spec
+	spec, err := a.dao.GetSpec(si.Spec.ID)
+	if err != nil {
+		// etcd return not found i.e. code 100
+		if client.IsKeyNotFound(err) {
+			return nil, ErrorNotFound
+		}
+		// otherwise unknown error bubble it up
+		return nil, err
+	}
+
+	// Make sure that we can find the plan we're updating from. This should probably never fail, but cover our tail.
+	var updateFromPlanKey = -1
+	for k, v := range spec.Plans {
+		if v.Name == (*si.Parameters)["_apb_plan_id"] {
+			updateFromPlanKey = k
+		}
+	}
+	if updateFromPlanKey == -1 {
+		a.log.Error("The plan %s, specified for updating from on instance %s, does not exist.", (*si.Parameters)["_apb_plan_id"], si.ID)
+		return nil, ErrorPlanNotFound
+	}
+	updateFromPlan := spec.Plans[updateFromPlanKey]
+
+	// Make sure we can find the plan we're updating to.
+	var updateToPlanKey = -1
+	for k, v := range spec.Plans {
+		if v.Name == req.PlanID {
+			updateToPlanKey = k
+		}
+	}
+	if updateToPlanKey == -1 {
+		a.log.Error("The plan %s, specified for updating to on instance %s, does not exist.", req.PlanID, si.ID)
+		return nil, ErrorPlanNotFound
+	}
+	updateToPlan := spec.Plans[updateFromPlanKey]
+
+	// Make sure that the current plan supports updating to the requested plan if a plan update was requested.
+	if req.PlanID != (*si.Parameters)["_apb_plan_id"] {
+		var updatable = false
+		for _, v := range updateFromPlan.UpdatesTo {
+			if v == req.PlanID {
+				updatable = true
+				(*si.Parameters)["_apb_plan_id"] = req.PlanID
+			}
+		}
+		if updatable == false {
+			a.log.Error("The current plan, %s, cannot be updated to the requested plan, %s.", (*si.Parameters)["_apb_plan_id"], req.PlanID)
+			return nil, ErrorPlanUpdateNotPossible
+		}
+	}
+
+	// Make sure that parameters requested for update exist on the new plan and that they are updatable.
+	for k, v := range req.Parameters {
+		var paramUpdated = false
+		for _, key := range updateToPlan.Parameters {
+			if key.Name == k {
+				if key.Updatable == false {
+					a.log.Error("Tried to update non-updatable parameter, %s, on instance %s.", k, si.ID)
+					return nil, ErrorParameterNotUpdatable
+				}
+				(*si.Parameters)[k] = v
+				paramUpdated = true
+			}
+		}
+		if paramUpdated == false {
+			a.log.Error("Parameter %s, requested for update on instance %s, does not exist.", k, si.ID)
+			return nil, ErrorParameterNotFound
+		}
+	}
+
+	// We're ready to provision so save
+	if err = a.dao.SetServiceInstance(instanceUUID.String(), si); err != nil {
```

Comment:
I think we might be doing this during normal provisions, but I'll raise the question: Do we want to be persisting service instances before we know they've *actually* been successfully created? Do they need to be extended to add some kind of `Status` field that's set to {Success|Error} at the conclusion of the APB run?

This is probably outside the scope of this PR, but worth consideration.